### ISSUE TITLE
Can line counting possibly be more intelligent?

### ISSUE THREAD

**Issue by lleolin** [bbatsov/rubocop#2686]
I have this:

```ruby
def update_sales_rep_attrs
  params
    .require(:sales_rep)
    .permit(
      :artist_id,
      :business_development_leader_id,
      :commission_override,
      :direct_deposit_status,
      :returning,
      :team_size,
      :year_sales_goal
    )
end
```
In my eyes, this should count as one line. It's a singular statement split up over several lines for readability and edit-ability (it's easier to delete a single line than remove something from a list). Sadly, Rubocop counts this method as 11 lines long.

In contrast, when I use SimpleCov, it counts that same statement as being a single "line", which gives a more accurate portrayal of the code's true length.

The line-length metric, in tandem with the method length metric, creates sort of a catch 22 where I'm forced to break code into multiple lines to keep it under 80 columns, but then sometimes "punished" for doing so because now all that text extends vertically instead of horizontally.

This can lead down somewhat-contrived paths where I try to define that list of attributes outside of a method so that Rubocop won't "care". For example, I could define those attributes as a constant somewhere, which is actually somewhat reasonable, but this is not typical Rails convention.

**Comment by alexdowad** [comment]
Limiting the line length of methods can be useful to keep a method all on one "screen", so you can see the whole definition at once. If that is what you care about, RC's method of counting lines is the best.

**Comment by jawshooah** [comment]
Also consider that, by SimpleCov's apparent definition of "line", you could have an arbitrarily long chain of methods spanning hundreds of *editor* lines treated as a single *lexical* line.

**Comment by lleolin** [comment]
That was an astoundingly fast reply.

I perceive the method length cop as a way to keep me from writing methods that do too much. Indeed, I have found that it has coerced me into breaking methods down into smaller and more functional pieces.

I agree that vertical readability is a legitimate concern too, although I would imagine that most people are working with a bit more than 10 lines. Similar to the 80 character width, it's not an entirely arbitrary number pulled from the sky, but it *is* a line drawn in the sand to enforce some kind of standardization.

I could of course change my .rubocop.yml to allow more lines, but the thing about attribute-whitelisting methods like the one I have is that they can be arbitrarily long, and I'd be essentially bumping the limit for the handful of methods that manage to go past 10 lines.

@jawshooah That's true, but the amount of chaining that can be done should be self-limiting in practice.

If nothing else, I would argue that my method body should be counted as 3 lines, with the list of params and the closing parentheses being counted as part of the `.permit` line rather than 8 individual lines.

**Comment by alexdowad** [comment]
Some people will prefer to have each line counted as a "line", some will prefer to have each statement counted as a... "line". If you are interested in counting statements and not lines, then presumably you would like to see `obj.a; obj.b; obj.c` counted as 3 "lines".

Either way, this cop will not be ideal for everyone. Unless the vast majority of users prefer the latter definition of "line", I say we should keep the current behavior.

To force people to break large, overly complex methods down, we have things like the ABC cop and perceived complexity cop. They do things like counting conditionals, method calls, etc.

**Comment by lleolin** [comment]
>Some people will prefer to have each line counted as a "line", some will prefer to have each statement counted as a... "line".

That seems to be an argument to make it at least configurable.

>If you are interested in counting statements and not lines, then presumably you would like to see obj.a; obj.b; obj.c counted as 3 "lines".

Yes. (That said, Rubocop disallows semicolons by default.)

>To force people to break large, overly complex methods down, we have things like the ABC cop and perceived complexity cop. They do things like counting conditionals, method calls, etc.

This is a fair point. Maybe I should just disable method length entirely for my purposes.

To be honest, I don't really see the point in limiting the method length to ten lines purely for aesthetic or readability reasons. The 80-column width is necessary because scrolling both horizontally and vertically when trying to read or edit code is not a fun experience, but having to scroll vertically a bit to figure out what code is doing is something we all inevitably have to do, and I imagine almost all of us can fit a good deal more than 10 lines on our screen at a time.

I try to run Rubocop using the default settings (maybe disabled for code I didn't write, like generated configs) because if I just disable cops that are convenient to disable, and say "that doesn't really matter", then I'm going down a slippery slope where I eschew compliance whenever it's easy. For me, it's kind of an all-or-nothing thing.

**Comment by alexdowad** [comment]
> That seems to be an argument to make it at least configurable.

Maintainers, would you accept a patch to make this configurable? Or a patch which adds a new cop?

**Comment by jonas054** [comment]
I'd like to see `MethodLength` remain as it is today. I've argued before that `MethodLength` should just measure the length of a method, and that length should be defined as number of (non-comment by default) lines. See #494, and also #1077 that disucusses a `MethodExpressionCount` cop that eventually was ditched in favor of `AbcSize`.

A new cop sounds better than changing `MethodLength`, but can we see an example where `AbcSize` isn't good enough and we need another metric?

**Comment by lleolin** [comment]
I figured this had been discussed before, but I didn't see those two issues when I did a search. (I think I searched for "line count" or some such.)

It makes sense not to change MethodLength; I misunderstood its purpose. AbcSize seems to do a good enough job at addressing my concerns. I guess the best thing for me to do is disable the MethodLength cop. I was also unaware that I could disable certain cops for a particular method using comments.
